File: solution/207. Course Schedule/207_02.cpp
class Solution {
public:
    bool canFinish(int n, vector<vector<int>>& prerequisites) {
        adj.resize(n);
        colors.resize(n);
        for (auto& p : prerequisites) {
            int a = p[0], b = p[1];
            adj[a].push_back(b);
        }
        
        for (int i = 0; i < n; ++i)
            if (!dfs(i))
                return false;
        return true;
    }
    
private:
    vector<vector<int>> adj;
    vector<int> colors; // 0: unknow, 1: visiting, 2: visited
    
    bool dfs(int node) {
        if (colors[node] == 1) return false;
        if (colors[node] == 2) return true;
        
        colors[node] = 1;
        for (int next : adj[node]) {
            if (!dfs(next))
                return false;
        }
        colors[node] = 2;
        return true;
    }
};

File: solution/207. Course Schedule/207_02.py
class Solution:
    def canFinish(self, n: int, prerequisites: List[List[int]]) -> bool:
        adj = [[] for _ in range(n)]
        colors = [0 for _ in range(n)] # 0: unknow, 1: visiting, 2: visited

        for a, b in prerequisites:
            adj[a].append(b)
        
        def dfs(node):
            if colors[node] == 1:
                return False
            if colors[node] == 2:
                return True
            colors[node] = 1
            for next in adj[node]:
                if not dfs(next):
                    return False
            colors[node] = 2
            return True
        
        for i in range(n):
            if not dfs(i):
                return False
        return True
